content/en-us/luau/type-checking.md

```diff
@@ -0,0 +1,352 @@
+---
+title: Luau Typing
+description: Luau uses gradual typing through the use of type annotations and inference.
+---
+
+Luau uses a typing engine which gives every variable an inferred type.
+
+```lua
+local foo = "bar" --> string
+local x = 5 --> number
+```
+
+## Defining a Type
+
+A type can be defined using the `type` keyword.
+
+```lua
+type Vector2 = {x: number, y: number}
+```
+
+## Inference Modes
+
+There are three type inference modes that Luau can use in `Class.Script|Scripts`:
+
+* `--!nocheck` - Don't check types
+* `--!nonstrict` - Default mode for all scripts, only asserts variable types if they are explicitly annotated
+* `--!strict` - Asserts all types based off the inferred or explicitly annotated type
+
+## Types
+
+A type annotation can be defined using the `:` operator after a variable, followed by a type definition. By default, in `nonstrict` mode, all variables are assigned the type `any`.
+
+```lua
+local foo: string = "bar"
+local x: number = 5
+```
+
+There are 4 primitive types that can be used in an annotation:
+* `nil` - no value
+* `boolean` - `true`/`false` value
+* `number` - a numeric value
+* `string` - text
+
+Functions and tables are defined differently, and within Roblox, all `Class.Instance|Instances` and `DataType.Enums` have their own types.
+
+```lua
+part.Touched:Connect(function(otherPart: BasePart)
+	local brickColor: Enum.BrickColor = otherPart.BrickColor
+end)
+```
+
+An optional type can be defined using `?` at the end of the annotation.
+
+```lua
+local foo: string? = nil
+```
+
+### Literal Types
+
+Strings and booleans can also be cast to literal values, instead of using `string` and `boolean` respectively.
+
+```lua
+local alwaysHelloWorld: "Hello World!" = "Hello World!"
+local alwaysTrue: true = false -- TypeError: Type 'false' could not be converted into 'true'
+
+alwaysHelloWorld = "Goodbye World!" -- TypeError: Type '"Goodbyte World"' could not be into '"Hello World!"'
+```
+
+### Type Casts
+
+Sometimes, you may need to assist the typechecker by changing the type of a value using type casts, this can be done with the `::` operator.
+
+```lua
+local numericValue = 1
+
+local value = numericValue :: any -- ok, all expressions may be cast to 'any'
+local flag = numericValue :: boolean -- not ok, invalid 'number' to 'boolean' conversion
+```
+
+## Function Typing
+
+Assume the following function:
+
+```lua
+local function add(x, y)
+	return x + y
+end
+```
+
+This function will add x to y, however it might error if a string is given to this function, Luau doesn't know that this function can only use numbers, lets rework this function such that it does constrain `x` and `y` to be numbers.

```

Comment:
```suggestion
This function adds `x` to `y`, but will error if a string is passed as an argument. Luau doesn't know that this function can only use numbers, so lets type this function to constrain `x` and `y` as numbers.
```

Maybe someone else could rephrase this better; it still sounds kinda weird to me.